### PR TITLE
JetPay V2: Add new gateway

### PR DESCRIPTION
**Pull request by shasum**
New adapter that talks to [JetPay Host XML Specification V2.03.01.pdf](https://github.com/activemerchant/active_merchant/files/1030427/JetPay.Host.XML.Specification.V2.03.01.pdf)

---

```
$ RUBYOPT=W0 be rake test:remote TEST=test/remote/gateways/remote_jetpay_v2_test.rb 
19 tests, 86 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
```

```
$ RUBYOPT=W0 be rake test:units TEST=test/unit/gateways/jetpay_v2_test.rb 
17 tests, 66 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
```

### CODE REVIEW

File: lib/active_merchant/billing/gateways/jetpay_v2.rb [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+      def build_xml_request(transaction_type, options = {}, transaction_id = nil, &block)
+        xml = Builder::XmlMarkup.new
+        xml.tag! 'JetPay', 'Version' => API_VERSION do
+          # Basic values needed for any request
+          xml.tag! 'TerminalID', @options[:login]
+          xml.tag! 'TransactionType', transaction_type
+          xml.tag! 'TransactionID', transaction_id.nil? ? generate_unique_id.slice(0, 18) : transaction_id
+          xml.tag! 'Origin', options[:origin] || 'INTERNET'
+          xml.tag! 'IndustryInfo', 'Type' => options[:industry_info] || 'ECOMMERCE'
+          xml.tag! 'Application', options[:application] || 'n/a', 'Version' => options[:application_version] || '1.0'
```
Comment (davidsantoso) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: I think this syntax will error out? ` 'n/a', 'Version' => options[:application_version]` both the comma after `n/a` and the `'Version' => options[...]`. 

I guess that makes me wonder if we can just simplify it with:

`options[:application] || '1.0`

Unless I'm missing something about the JetPay v2 API where we might need to send something related to that middle value.

Comment (davidsantoso) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: I should note this goes for the line directly beneath this one as well.

File: lib/active_merchant/billing/gateways/jetpay_v2.rb [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+      end
+
+      def supports_scrubbing
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r((>)\d+(</CardNum>)), '\1[FILTERED]\2').
```
Comment (davidsantoso) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: Any particular reason to not have this be: %r((<CardNum>)\d+(</CardNum>)?

File: lib/active_merchant/billing/gateways/jetpay_v2.rb (hunk 1 above)
Comment (shasum) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: @davidsantoso 

Based off of the XML builder incantation `xml.tag! <name>, <value>, <attributes hash>`

the following

```
xml.tag! 'Application', options[:application] || 'n/a', 'Version' => options[:application_version] || '1.0'
```
helps output a markup such as

`<Application Version=1.0>spreedly</Application>`


The commas and || could be bit misleading but essentially it's 3 comma separated arguments with defaults for argument 2 and 3.

Couldn't find a better ref: https://apidock.com/rails/Builder/XmlMarkup but it talks about `tag! will also take text and attribute arguments (after the tag name) like normal markup methods`.

File: lib/active_merchant/billing/gateways/jetpay_v2.rb (hunk 2 above)
Comment (shasum) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: @davidsantoso Reason is to target numbers that fall within `CardNum` and `CVV2` tags only. Not including tag names as part of the pattern match may scrub off things like amount and order number.

Comment (davidsantoso) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: @shasum so sorry- I just realized that I forgot to add tick marks around my code and GitHub escaped the contents 😞 See my updated question!

File: lib/active_merchant/billing/gateways/jetpay_v2.rb (hunk 1 above)
Comment (davidsantoso) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: @shasum ah I see. So really it's this-

```
xml.tag! 'Application', (options[:application] || 'n/a'), ('Version' => options[:application_version] || '1.0')
```

Comment (shasum) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: Yep, the parens definitely helps :)

File: lib/active_merchant/billing/gateways/jetpay_v2.rb (hunk 2 above)
Comment (shasum) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: @davidsantoso The start tag encompasses couple attributes.

`<CardNum CardPresent="false" Tokenize="true">4xxx3xxx2xxx1xxx</CardNum>`
 
Hence the focus on `>4xxx3xxx2xxx1xxx</CardNum>`

File: lib/active_merchant/billing/gateways/jetpay_v2.rb [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        commit(money, build_capture_request(transaction_id, money, options), token)
+      end
+
+      def void(reference, options = {})
+        transaction_id, approval, amount, token = reference.split(";")
+        commit(amount.to_i, build_void_request(amount.to_i, transaction_id, approval, token, options), token)
+      end
+
+      def credit(money, credit_card, options = {})
+        commit(money, build_credit_request('CREDIT', money, nil, credit_card, nil, options))
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: Slight inconsistency, unless there's some logic nuance: 'CREDIT' is sent from here to the `build_redit_request` for both `refund` and general `credit` actions, but the transaction type for other actions is sent from their `build_` methods.

Comment (curiousepic) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: Otherwise, LGTM!

Comment (shasum) [inline lib/active_merchant/billing/gateways/jetpay_v2.rb]: Good catch. I'll edit that plus the parens before merging.

**Pull request merged**